power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/dataset_handler_fwd.hpp

```diff
@@ -0,0 +1,41 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+
+// forward declarations for handling dataset and buffer related stuff
+
+#include <concepts>
+
+namespace power_grid_model::meta_data {
+
+struct data_mutable_t {};
+struct data_immutable_t {};
+
+template <typename T>
+concept data_mutable_tag = std::same_as<T, data_mutable_t> || std::same_as<T, data_immutable_t>;
+template <data_mutable_tag T> constexpr bool is_data_mutable = std::same_as<T, data_mutable_t>;
+
+static_assert(data_mutable_tag<data_mutable_t>);
+static_assert(data_mutable_tag<data_immutable_t>);
+static_assert(is_data_mutable<data_mutable_t>);
+static_assert(!is_data_mutable<data_immutable_t>);
+
+struct indptr_mutable_t {};
+struct indptr_immutable_t {};
+
+template <typename T>
+concept indptr_mutable_tag = std::same_as<T, indptr_mutable_t> || std::same_as<T, indptr_immutable_t>;
+template <indptr_mutable_tag T> constexpr bool is_indptr_mutable = std::same_as<T, indptr_mutable_t>;
+
+static_assert(indptr_mutable_tag<indptr_mutable_t>);
+static_assert(indptr_mutable_tag<indptr_immutable_t>);
+static_assert(is_indptr_mutable<indptr_mutable_t>);
+static_assert(!is_indptr_mutable<indptr_immutable_t>);
+
+template <data_mutable_tag data_mutable, indptr_mutable_tag indptr_mutable>
+    requires(is_data_mutable<data_mutable> || !is_indptr_mutable<indptr_mutable>)
+class DatasetHandler;
```

Comment:
alternatively, I could create 3 tags:

```suggestion
struct const_dataset_t {};   // may be merged with the definition from dataset.hpp
struct mutable_dataset_t {}; // may be merged with the definition from dataset.hpp
struct writable_dataset_t{};

// to make the distinction between dataset handler tags and dataset tags
template <typename T>
concept dataset_handler_tag = std::same_as<T, const_dataset_t> || std::same_as<T, mutable_dataset_t> || std::same_as<T, writable_dataset_t>;

template <dataset_handler_tag T> constexpr bool is_data_mutable = std::same_as<T, mutable_dataset_t> || std::same_as<T, writable_dataset_t>;
template <dataset_handler_tag T> constexpr bool is_indptr_mutable = std::same_as<T, writable_dataset_t>;

static_assert(dataset_handler_tag<const_dataset_t>);
static_assert(dataset_handler_tag<mutable_dataset_t>);
static_assert(dataset_handler_tag<writable_dataset_t>);
static_assert(!is_data_mutable<const_dataset_t>);
static_assert(is_data_mutable<mutable_dataset_t>);
static_assert(is_data_mutable<writable_dataset_t>);
static_assert(!is_indptr_mutable<const_dataset_t>);
static_assert(!is_indptr_mutable<mutable_dataset_t>);
static_assert(is_indptr_mutable<writable_dataset_t>);

template <dataset_handler_tag dataset_handler_type>
class DatasetHandler;
```

The three types defined in `dataset_handler.hpp` would then become

```cpp
using ConstDatasetHandler = DatasetHandler<const_dataset_t>;
using MutableDatasetHandler = DatasetHandler<mutable_dataset_t>;
using WritableDatasetHandler = DatasetHandler<writable_dataset_t>;
```

Reply:
I prefer the alternative way. Using three tags. The original implementation has two template arguments has to do that `bool` only has two states.